frontend/lib/resolution/resolution-queries.cpp

```diff
@@ -3837,23 +3837,51 @@ const TypedFnSignature* tryResolveInitEq(Context* context,
   return c.mostSpecific().only().fn();
 }
 
-static const TypedFnSignature* tryResolveAssignHelper(
-    Context* context, const uast::AstNode* astForScopeOrErr,
-    const types::Type* t, bool asMethod) {
-  auto lhsType = QualifiedType(QualifiedType::CONST_REF, t);
-  auto rhsType = QualifiedType(QualifiedType::CONST_REF, t);
+const TypedFnSignature* tryResolveDeinit(Context* context,
+                                         const AstNode* astForScopeOrErr,
+                                         const types::Type* t,
+                                         const PoiScope* poiScope) {
+  if (!t->getCompositeType()) return nullptr;
+
+  QualifiedType qt(QualifiedType::VAR, t);
+
+  std::vector<CallInfoActual> actuals;
+  actuals.push_back(CallInfoActual(qt, USTR("this")));
+  auto ci = CallInfo(/* name */ USTR("deinit"),
+                     /* calledType */ qt,
+                     /* isMethodCall */ true,
+                     /* hasQuestionArg */ false,
+                     /* isParenless */ false,
+                     actuals);
+
+  const Scope* scope = nullptr;
+  if (astForScopeOrErr) scope = scopeForId(context, astForScopeOrErr->id());
+
+  auto c = resolveGeneratedCall(context, astForScopeOrErr, ci, scope, poiScope);
+  return c.mostSpecific().only().fn();
+}
+
+static const TypedFnSignature*
+tryResolveAssignHelper(Context* context,
+                       const uast::AstNode* astForScopeOrErr,
+                       const types::Type* lhsType,
+                       const types::Type* rhsType,
+                       bool asMethod) {
+  auto qtLhs = QualifiedType(QualifiedType::CONST_REF, lhsType);
+  auto qtRhs = QualifiedType(QualifiedType::CONST_REF, rhsType);
```

Comment:
`owned` will have an `=` that mutates the RHS. I don't htink that's relevant here, so you probably want `QualifiedType::VAR` here.